jobs: update jobs and job_info join queries

adityamaru opened the pull request:
This change updates the two places where we perform
a join over the `system.jobs` and `system.job_info`
table to use more performant and correct queries.

The queries are more performant because of an additional
predicate on the `job_id` in each of the CTE queries. This
prevents a full table scan when we are only concerned with
rows corresponding to a particular job. This has been verified
by running both variants of the query through the EXPLAIN ANALYZE (plan)
tool. This tool occassionally reports a full table scan in the CTE queries
when there is no predicate involved.

The queries are more correct because they take into consideration
a job having >1 row for a particular info_key. This is possible
during upgrades. Jobs start writing their payload and progress to
the job_info table once V23_1CreateSystemJobInfoTable is active.
If a job is still running when the cluster steps through
V23_1JobInfoTableIsBackfilled then the job will have duplicate entries
for the payload and prgress rows. This is because we blindly backfill
the contents of the jobs table into the job_info table without
clearing previous entries corresponding to the job.

This change also adds a red-green regression test for the above
correctness problem.

Fixes: https://github.com/cockroachdb/cockroach/issues/99101
Informs: #99219
Release note: None
Epic: CRDB-8964

miretskiy reviewed (approved):
Do you think we should have additional tests? This is 2nd time this query being changed.

adityamaru committed:
jobs: update jobs and job_info join queries

This change updates the two places where we perform
a join over the `system.jobs` and `system.job_info`
table to use more performant and correct queries.

The queries are more performant because of an additional
predicate on the `job_id` in each of the CTE queries. This
prevents a full table scan when we are only concerned with
rows corresponding to a particular job. This has been verified
by running both variants of the query through the EXPLAIN ANALYZE (plan)
tool. This tool occassionally reports a full table scan in the CTE queries
when there is no predicate involved.

The queries are more correct because they take into consideration
a job having >1 row for a particular info_key. This is possible
during upgrades. Jobs start writing their payload and progress to
the job_info table once V23_1CreateSystemJobInfoTable is active.
If a job is still running when the cluster steps through
V23_1JobInfoTableIsBackfilled then the job will have duplicate entries
for the payload and prgress rows. This is because we blindly backfill
the contents of the jobs table into the job_info table without
clearing previous entries corresponding to the job.

This change also adds a red-green regression test for the above
correctness problem.

Fixes: #99101
Release note: None

adityamaru commented:
@knz / @miretskiy the scope of this change has expanded a bit to also make the queries avoid full table scans. Can you give it another look please?

dt commented on pkg/sql/crdb_internal.go:
```diff
@@ -903,6 +903,22 @@ WITH
+	// queries.
+	systemJobsAndJobInfoBaseQueryWithIDPredicate = `
+WITH
+	latestpayload AS (SELECT job_id, value FROM system.job_info AS payload WHERE info_key = 'legacy_payload'::BYTES AND job_id = $1),
```
Why don't these two get the same order/limit clauses?

adityamaru replied:
rah SQL is hard. I should give them the order clause. I didn't give them the limit clause because the driver uses a row iterator so it _can_ handle more than one result. The caller of `update.go` used `QueryRow` and so it would panic if it saw more than one result. 

That said I now realize this means 4 rows in `SHOW JOBS` for the same job_id which is no good. Is it weird that `SELECT * from crdb_internal.system_jobs WHERE id = $1` doesn't show all the rows corresponding to id $1?

adityamaru replied:
@dt this was a good call out, even when we do not have a predicate we need to ensure we only display a single row per job. To make this happen I've added a `distinct` clause to the `id` column along with the `ORDER BY` clauses to the CTEs. I also added some more checks to the test.

adityamaru commented on pkg/jobs/update.go:
```diff
@@ -379,8 +379,8 @@ func getSelectStmtForJobUpdate(
 WITH
-	latestpayload AS (SELECT job_id, value FROM system.job_info AS payload WHERE info_key = 'legacy_payload'::BYTES),
-	latestprogress AS (SELECT job_id, value FROM system.job_info AS progress WHERE info_key = 'legacy_progress'::BYTES)
+	latestpayload AS (SELECT job_id, value FROM system.job_info AS payload WHERE info_key = 'legacy_payload'::BYTES AND job_id = $1 ORDER BY written DESC LIMIT 1),
```
This is strange. `TestBackupTenantImportingTable` doesn't like the LIMIT 1 clause at the end of this query. After the restore in that test when we restart the tenant the test immediately fails with:

```
E230328 23:51:18.245401 6288 2@rpc/context.go:2388  [T10,nsql2,rnode=1,raddr=127.0.0.1:62608,class=default,rpc] 335  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: co
nnection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
I230328 23:51:18.245434 6288 rpc/context.go:2302  [T10,nsql2,rnode=1,raddr=127.0.0.1:62608,class=default,rpc] 336  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection er
ror: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
I230328 23:51:18.245467 6858 1@circuitbreaker/circuitbreaker.go:322  [T10,nsql2] 337  circuitbreaker: rpc 127.0.0.1:62620 [n1] tripped: failed to connect to n1 at 127.0.0.1:62608: initial connection heartbeat fail
ed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
I230328 23:51:18.245503 6858 1@circuitbreaker/circuitbreaker.go:447  [T10,nsql2] 338  circuitbreaker: rpc 127.0.0.1:62620 [n1] event: BreakerTripped
W230328 23:51:18.245473 6514 2@rpc/nodedialer/nodedialer.go:196  [T10,nsql2,client=127.0.0.1:62626,user=root,intExec=select-job] 339  unable to connect to n1: failed to connect to n1 at 127.0.0.1:62608: initial co
nnection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
W230328 23:51:18.245581 6858 sql/colflow/colrpc/outbox.go:183  [T10,nsql2,client=127.0.0.1:62626,user=root,intExec=select-job,f39c33432,distsql.stmt=WITH latestpayload AS (SELECT job_id, value FROM system.job_info
 AS payload WHERE (info_key = '_'::BYTES) AND (job_id = $1) ORDER BY written DESC LIMIT _), latestprogress AS (SELECT job_id, value FROM system.job_info AS progress WHERE (info_key = '_'::BYTES) AND (job_id = $1)
ORDER BY written DESC LIMIT _) SELECT status, payload.value AS payload, progress.value AS progress, COALESCE(last_run, created), COALESCE(num_runs, _) FROM system.jobs AS j INNER JOIN latestpayload AS payload ON j
.id = payl,distsql.gateway=2,distsql.appname=$ internal-select-job,distsql.txn=31defc85-a5e6-4cdc-97ca-ed6431fd2104,streamID=6] 340  Outbox Dial connection error, distributed query will fail: failed to connect to
n1 at 127.0.0.1:62608: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:62608: connect: connection refused" [code 14/Unavailable]
```

Dropping the LIMIT 1 "fixes" it 🤔

adityamaru replied:
I would be lying if I said I knew what was exactly going on but pushing the CANCEL into the succeeds soon allows the test to succeed. Why a LIMIT 1 clause in a CTE would so reliably make the CANCEL fail is still a mystery. This test does stop the tenant server, and restart it post-restore so it's possible this is related to the various slow quiesce issues we're seeing elsewhere.

adityamaru replied:
I'm actually going to go out on a limb and say whatever Yahor finds https://github.com/cockroachdb/cockroach/issues/99753 is going to be related. Pattern matching `Outbox Dial connection error, distributed query will fail: dial: context canceled` to that issue

knz reviewed (commented):
> Reviewed 1 of 4 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99659)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, and @miretskiy)

___
*[`pkg/jobs/update.go` line 382 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/99659#-NRdYwhfv8KqSWHqWVKW-r2-382:-NRgIMdWBO8SEwcrr-wq:bmgsooz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c072906e8bbe05cd284c723cb1517dee4e5b0de6/pkg/jobs/update.go#L382)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I'm actually going to go out on a limb and say whatever Yahor has found in https://github.com/cockroachdb/cockroach/issues/99753 is probably going to be related 🤞  Pattern matching `Outbox Dial connection error, distributed query will fail: dial: context canceled` to that issue
</blockquote></details>

Did you ask Yahor to look at this PR? The fact you are able to so reliably reproduce the issue here could help Yahor troubleshoot the problem more effectively.
___
*[`pkg/sql/crdb_internal.go` line 900 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/99659#-NRgHzPfELjo-V9S1y0k:-NRgHzPfELjo-V9S1y0l:b-qw9pzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d05140f4be0f98f82496bd09e62f83bc62d80da/pkg/sql/crdb_internal.go#L900)):*
> ```Go
> 	latestprogress AS (SELECT job_id, value FROM system.job_info AS progress WHERE info_key = 'legacy_progress'::BYTES ORDER BY written DESC)
> 	SELECT 
> 		distinct(id), status, created, payload.value AS payload, progress.value AS progress,
> ```

nit: distinct is a keyword -> uppercase DISTINCT


<!-- Sent from Reviewable.io -->

adityamaru commented:
CI failure is https://github.com/cockroachdb/cockroach/pull/100099

knz commented:
you'll need to rebase this

adityamaru commented:
TFTR!

bors r+
